subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java

```diff
@@ -44,24 +42,20 @@
 import java.nio.file.Path;
 import java.util.HashSet;
 import java.util.Set;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
 
-/**
- * A {@link org.gradle.internal.vfs.VirtualFileSystem} which uses watches to maintain
- * its contents.
- */
-public class WatchingVirtualFileSystem extends AbstractDelegatingVirtualFileSystem implements WatchingAwareVirtualFileSystem, Closeable {
-    private static final Logger LOGGER = LoggerFactory.getLogger(WatchingVirtualFileSystem.class);
+public class DefaultFileSystemWatchingHandler implements FileSystemWatchingHandler, Closeable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFileSystemWatchingHandler.class);
     private static final String FILE_WATCHING_ERROR_MESSAGE_DURING_BUILD = "Unable to watch the file system for changes";
     private static final String FILE_WATCHING_ERROR_MESSAGE_AT_END_OF_BUILD = "Gradle was unable to watch the file system for changes";
 
     private final FileWatcherRegistryFactory watcherRegistryFactory;
+    private final AbstractVirtualFileSystem virtualFileSystem;
```

Comment:
There is a code smell here... `AbstractVirtualFileSystem` should be an interface instead. Maybe what we need is a `VFS.update(UpdateFunction)` method.

Reply:
That sounds like `VirtualFileSystem` becoming an `AtomicSnapshotHierarchyReference`. That isn't necessarily bad. We also could separate the `AtomicSnapshotHierarchyReference` from the VFS making it its own first class object. The snapshot hierarchy is the data in the VFS, while `VirtualFileSystem` allows querying the actual file system, using the backing `SnapshotHierarchy`.

`FileSystemWatchingHandler` operates only on the snapshot hierarchy, publishing changes to it to the watcher registry.

I'll try something out. Let's look at this together tomorrow as well.

Reply:
`FileSystemWatchingHandler` now only uses the `AtomicSnapshotHierarchyReference` and `AbstractVirtualFileSystem` is gone.